File: be-high-level/router.js
import express from "express";
import { getAllTokens, addToken, removeToken, clearDB, tryLuck, getTotalWins } from "./utils/data.js";
import { generateNewToken, getUserToken } from "./utils/token.js";

const router = express.Router();

router.get('/', (_, res) => {
    res.send('Server is working');
})

router.get('/all', async (req, res) => {
    try {
        const tokens = await getAllTokens();
        res.status(200).send(tokens);
    } catch (error) {
        console.error("Error fetching tokens:", error);
        res.status(500).send("Internal Server Error");
    }
});

router.post('/api/login', async (req, res) => {
    try {
        const randomToken = generateNewToken(req);
        if (randomToken) {
            const success = await addToken(randomToken);
            if (success) {
                res.send({ token: randomToken });
            } else {
                res.status(500).send("Error saving token");
            }
        } else {
            res.status(401).send({ "error": "Wrong email and password" });
        }
    } catch (error) {
        console.log("Error in /login route:", error);
        res.status(500).send("Server error");
    }
})

router.post('/api/logout', async (req, res) => {
    try {
        const userToken = getUserToken(req);
        const success = await removeToken(userToken);
        if (success) {
            res.send(`"OK"`);
        } else {
            res.status(500).send("Error removing token");
        }
    } catch (error) {
        console.log("Error in /logout route:", error);
        res.status(500).send("Server error");
    }
})

router.post('/api/try_luck', async (req, res) => {
    try {
        const userToken = getUserToken(req);
        const totalWins = await getTotalWins();
        const didWin = totalWins > 30 ? Math.random() < 0.4 : Math.random() < 0.7
        const success = await tryLuck(userToken, didWin)
        if (success) {
            res.send({ "win": didWin })
        } else {
            res.status(500).send("Error trying");
        }
    } catch (error) {

    }
})

router.get('/total_wins', async (req, res) => {
    try {
        const totalWins = await getTotalWins();
        res.send({ totalWins })
    } catch (error) {

    }
})

router.delete('/', async (req, res) => {
    try {
        const success = await clearDB();
        if (success) {
            res.send("CLEARED");
        } else {
            res.status(500).send("Error clearing DB");
        }
    } catch (error) {
        console.log("Error in delete route:", error);
        res.status(500).send("Server error");
    }
})

export default router;

File: be-low-level/src/router.rs
use actix_web::{web, HttpResponse, HttpRequest, Responder, Error};
use crate::utils::data::*;
use crate::utils::token::*;

pub fn router(cfg: &mut web::ServiceConfig) {
    cfg.service(
        web::resource("/")
            .route(web::get().to(|| async { HttpResponse::Ok().body("Server is working") })),
    )
    .service(web::resource("/api/login").route(web::post().to(login)))
    .service(web::resource("/api/logout").route(web::post().to(logout)))
    .service(web::resource("/api/try_luck").route(web::post().to(try_luck_user)));
}

async fn login(req: web::Json<LoginRequest>) -> Result<impl Responder, Error> {
    match generate_new_token(req.into_inner()) {
        Some(random_token) => {
            if add_token(random_token.clone().as_str()).await {
                Ok(HttpResponse::Ok().json(LoginResponse { token: random_token }))
            } else {
                Ok(HttpResponse::InternalServerError().body("Error saving token"))
            }
        }
        None => Ok(HttpResponse::Unauthorized().json(ErrorResponse { error: "Wrong email and password".to_string() })),
    }
}


async fn logout(req: HttpRequest) -> Result<impl Responder, Error> {
    let user_token = get_user_token(&req);
    if remove_token(&user_token).await {
    Ok(HttpResponse::Ok().body("\"OK\""))
    } else {
        Ok(HttpResponse::InternalServerError().body("Error removing token"))
    }
}

async fn try_luck_user(req: HttpRequest) -> Result<impl Responder, Error> {
    let user_token = get_user_token(&req);
    let total_wins = get_total_wins().await.unwrap_or(0);
    let did_win = if total_wins > 30 { rand::random::<f32>() < 0.4 } else { rand::random::<f32>() < 0.7 };
    
    if update_win(&user_token, did_win).await {
        Ok(HttpResponse::Ok().json(WinnerResponse { win: did_win }))
    } else {
        Ok(HttpResponse::InternalServerError().body("Error trying"))
    }
}